File: scripts/gc_translation/bible.py
"""Bible reference parsing, VI1934 lookup, and [[BIBLE:...]] sentinel resolution."""
from __future__ import annotations

import json
import re
from pathlib import Path

import yaml

from scripts.gc_translation.paths import BIBLE_REFS_PATH, VI1934_PATH

SENTINEL_RE = re.compile(
    r"(?P<lp>\()?\[\[BIBLE:(?P<ref>[^\]]+)\]\](?P<rp>\))?[.,;:!?]?"
)
REF_RE = re.compile(
    r"^\s*((?:[1-3]\s+)?[A-Za-z][A-Za-z. ]*?)\s+(\d+):(\d+)(?:-(\d+))?\s*$"
)
MULTI_REF_RE = re.compile(
    r"^\s*((?:[1-3]\s+)?[A-Za-z][A-Za-z. ]*?)\s+(\d+)(?::([\d,\-\s]+?))?\s*$"
)
PART_RE = re.compile(r"^(\d+)(?:-(\d+))?$")

SINGLE_CHAPTER_BOOKS = {"Obadiah", "Philemon", "2 John", "3 John", "Jude"}


def load_bible() -> dict[str, str]:
    return json.loads(Path(VI1934_PATH).read_text(encoding="utf-8"))


def load_bible_refs() -> dict[str, str]:
    return yaml.safe_load(Path(BIBLE_REFS_PATH).read_text(encoding="utf-8"))


def _canonicalize_book(bible_refs: dict[str, str], name: str) -> str:
    name = name.strip().rstrip(".")
    if name in bible_refs:
        vn = bible_refs[name]
        candidates = [e for e, v in bible_refs.items() if v == vn]
        return max(candidates, key=len)
    return name


def parse_ref(ref: str) -> tuple[str, int, int, int]:
    m = REF_RE.match(ref)
    if not m:
        raise ValueError(f"Cannot parse bible reference: {ref!r}")
    book_raw, chapter, vstart, vend = m.groups()
    bible_refs = load_bible_refs()
    book = _canonicalize_book(bible_refs, book_raw)
    return book, int(chapter), int(vstart), int(vend) if vend else int(vstart)


def parse_ref_multi(ref: str) -> tuple[str, int, list[tuple[int, int]], str]:
    """Parse a bible ref that may include comma-separated verse parts.

    Returns (book, chapter, parts, verse_spec) where parts is a list of
    (start, end) ranges and verse_spec is the normalized original verse string
    (e.g. "9,21-22") for use in the citation.
    """
    bible_refs = load_bible_refs()
    sc_re = re.compile(
        r"^\s*((?:[1-3]\s+)?[A-Za-z][A-Za-z. ]*?)\s+([\d,\-\s]+?)\s*$"
    )
    sc_m = sc_re.match(ref)
    if sc_m:
        book_try = _canonicalize_book(bible_refs, sc_m.group(1))
        if book_try in SINGLE_CHAPTER_BOOKS:
            ref = f"{book_try} 1:{sc_m.group(2)}"
    m = MULTI_REF_RE.match(ref)
    if not m:
        raise ValueError(f"Cannot parse bible reference: {ref!r}")
    book_raw, chapter, verse_part = m.groups()
    book = _canonicalize_book(bible_refs, book_raw)
    if verse_part is None:
        return book, int(chapter), [], ""
    parts: list[tuple[int, int]] = []
    normalized_chunks: list[str] = []
    for chunk in verse_part.split(","):
        chunk = chunk.strip()
        pm = PART_RE.match(chunk)
        if not pm:
            raise ValueError(f"Cannot parse verse part: {chunk!r}")
        a, b = pm.groups()
        start = int(a)
        end = int(b) if b else start
        parts.append((start, end))
        normalized_chunks.append(chunk)
    return book, int(chapter), parts, ",".join(normalized_chunks)


def lookup_verse(
    bible: dict[str, str], book: str, chapter: int, vstart: int, vend: int
) -> str | None:
    parts: list[str] = []
    for v in range(vstart, vend + 1):
        text = bible.get(f"{book} {chapter}:{v}")
        if text is None:
            return None
        parts.append(text)
    return " ".join(parts)


def vietnamese_book_name(bible_refs: dict[str, str], english: str) -> str:
    english = english.strip().rstrip(".")
    return bible_refs.get(english, english)


def resolve_sentinels(
    text: str,
    bible: dict[str, str] | None = None,
    bible_refs: dict[str, str] | None = None,
) -> tuple[str, list[str]]:
    if bible is None:
        bible = load_bible()
    if bible_refs is None:
        bible_refs = load_bible_refs()

    unresolved: list[str] = []

    def replace(m: re.Match[str]) -> str:
        ref_str = m.group("ref").strip()
        lp, rp = m.group("lp"), m.group("rp")
        # If unbalanced parens, restore them around the original on failure.
        unbalanced = bool(lp) ^ bool(rp)
        try:
            book, chapter, parts, verse_spec = parse_ref_multi(ref_str)
        except ValueError:
            unresolved.append(ref_str)
            return m.group(0)
        if not parts:
            verses: list[int] = []
            v = 1
            while bible.get(f"{book} {chapter}:{v}") is not None:
                verses.append(v)
                v += 1
            if not verses:
                unresolved.append(ref_str)
                return m.group(0)
            parts = [(verses[0], verses[-1])]
            verse_spec = f"{verses[0]}-{verses[-1]}"
        segments: list[str] = []
        for vstart, vend in parts:
            verse = lookup_verse(bible, book, chapter, vstart, vend)
            if verse is None:
                unresolved.append(ref_str)
                return m.group(0)
            segments.append(verse)
        vi_book = vietnamese_book_name(bible_refs, book)
        body = " ".join(segments)
        if book in SINGLE_CHAPTER_BOOKS:
            cite = f"{vi_book} {verse_spec}"
        else:
            cite = f"{vi_book} {chapter}:{verse_spec}"
        return f"\n\n> \"{body}\"\n> <cite>({cite})</cite>\n\n"

    return SENTINEL_RE.sub(replace, text), unresolved

File: scripts/egw_translate/lib/bible.ts
import { parseYaml } from "./yaml";

const SINGLE_CHAPTER_BOOKS = new Set(["Obadiah", "Philemon", "2 John", "3 John", "Jude"]);

export interface BibleLookup {
  bookNames: Map<string, string>;       // English → Vietnamese
  verses: Map<string, string>;          // "Book Ch:Vs" → text
}

export async function loadBibleLookup(refsPath: string, versesPath: string): Promise<BibleLookup> {
  const refsText = await Bun.file(refsPath).text();
  const parsedRefs = parseYaml(refsText);
  const bookNames = new Map<string, string>();
  for (const [en, vi] of Object.entries(parsedRefs)) {
    if (typeof vi === "string") bookNames.set(en, vi);
  }
  const versesJson = JSON.parse(await Bun.file(versesPath).text()) as Record<string, string>;
  const verses = new Map<string, string>(Object.entries(versesJson));
  return { bookNames, verses };
}

/** One chapter and its verse spans, e.g. { chapter: 8, spans: [[28, 28]] }. */
export interface RefSegment {
  chapter: number;
  spans: Array<[number, number]>;
}

export interface ParsedRef {
  canonicalBook: string;                // key used for vi1934.json lookup (e.g., "Matthew")
  viBook: string;                       // Vietnamese display name (e.g., "Ma-thi-ơ")
  chapter: number;                      // first segment's chapter (back-compat)
  verseSpans: Array<[number, number]>;  // first segment's spans (back-compat)
  segments: RefSegment[];               // all chapter/verse segments, in order
  display: string;                      // reproduced ref for the cite line, e.g., "24:20-25" or "8:28,6:57"
}

/**
 * Find the canonical English book key for a name. Tries exact, then case-insensitive.
 */
function findBookKey(raw: string, bookNames: Map<string, string>): string | null {
  if (bookNames.has(raw)) return raw;
  const lower = raw.toLowerCase();
  for (const key of bookNames.keys()) {
    if (key.toLowerCase() === lower) return key;
  }
  return null;
}

/**
 * Parse a sentinel body like "Matthew 24:20", "Luke 15:1,4", "Jude 3",
 * "Matt 24:20-25", or a multi-chapter list "John 8:28,6:57,8:50,7:18".
 *
 * Within the verse spec, comma-separated groups normally share the chapter from
 * the first group (e.g. "15:1,4" → 15:1 and 15:4). A group carrying its own colon
 * (e.g. "8:28,6:57") starts a new chapter for the verses that follow.
 */
export function parseRef(body: string, lookup: BibleLookup): ParsedRef | null {
  // Allow each comma group to optionally carry its own "chapter:" prefix.
  const m = body.trim().match(/^(\d?\s?[A-Za-z][A-Za-z. ]*?)\s+(\d+(?::\d+(?:-\d+)?)?(?:,(?:\d+:)?\d+(?:-\d+)?)*)$/);
  if (!m) return null;
  const rawBook = m[1].replace(/\s+/g, " ").trim();
  const spec = m[2].trim();

  const bookKey = findBookKey(rawBook, lookup.bookNames);
  if (!bookKey) return null;

  // Follow aliases: bible-refs.yaml maps both "Matt" and "Matthew" to "Ma-thi-ơ".
  // The canonical English book key (used for vi1934.json lookup) is the longest English
  // form pointing to the same Vietnamese name.
  const viName = lookup.bookNames.get(bookKey)!;
  let canonical = bookKey;
  for (const [en, vi] of lookup.bookNames) {
    if (vi === viName && en.length > canonical.length) canonical = en;
  }

  const segments: RefSegment[] = [];

  if (spec.includes(":")) {
    // Walk comma groups, carrying the chapter forward until a group overrides it.
    let chapter = 0;
    for (const group of spec.split(",")) {
      let versePart = group;
      if (group.includes(":")) {
        const [chStr, vStr] = group.split(":");
        chapter = Number(chStr);
        versePart = vStr;
      }
      if (chapter === 0) return null; // first group must establish a chapter
      const spans = parseVerseSpec(versePart);
      const last = segments[segments.length - 1];
      if (last && last.chapter === chapter) last.spans.push(...spans);
      else segments.push({ chapter, spans });
    }
  } else {
    if (!SINGLE_CHAPTER_BOOKS.has(canonical)) return null;
    segments.push({ chapter: 1, spans: parseVerseSpec(spec) });
  }

  const first = segments[0];
  return {
    canonicalBook: canonical,
    viBook: viName,
    chapter: first.chapter,
    verseSpans: first.spans,
    segments,
    display: spec,
  };
}

function parseVerseSpec(s: string): Array<[number, number]> {
  const parts = s.split(",");
  const out: Array<[number, number]> = [];
  for (const p of parts) {
    if (p.includes("-")) {
      const [a, b] = p.split("-").map(Number);
      out.push([a, b]);
    } else {
      const n = Number(p);
      out.push([n, n]);
    }
  }
  return out;
}

/**
 * Resolve a parsed ref to the concatenated Vietnamese verse text.
 *
 * Individual verses absent from VI1934 are skipped — some verses (e.g. Mark 9:44,
 * Matthew 17:21) are genuinely omitted from the 1934 text, and a range that spans
 * such a gap should still resolve from the verses that ARE present. Returns null
 * only when the reference resolves to no verses at all (a truly bad reference).
 */
export function lookupVerses(ref: ParsedRef, lookup: BibleLookup): string | null {
  const parts: string[] = [];
  for (const seg of ref.segments) {
    for (const [from, to] of seg.spans) {
      for (let v = from; v <= to; v++) {
        const key = `${ref.canonicalBook} ${seg.chapter}:${v}`;
        const text = lookup.verses.get(key);
        if (text != null) parts.push(text);
      }
    }
  }
  return parts.length ? parts.join(" ") : null;
}

/**
 * Replace every [[BIBLE:...]] sentinel in `text` with a formatted blockquote.
 * Unresolved sentinels are left in place; the caller can grep for them.
 */
export function resolveBibleSentinels(text: string, lookup: BibleLookup): { output: string; unresolved: string[] } {
  const unresolved: string[] = [];
  // Mirror gc_translation/bible.py SENTINEL_RE: consume optional wrapping parens
  // and one trailing punctuation char so they don't get orphaned when the sentinel
  // is replaced by an isolated blockquote paragraph.
  const replaced = text.replace(/[ \t]*\(?\[\[BIBLE:([^\]]+)\]\]\)?[.,;:!?]?[ \t]*/g, (match, body: string) => {
    const ref = parseRef(body, lookup);
    if (!ref) {
      unresolved.push(body.trim());
      return match;
    }
    const verseText = lookupVerses(ref, lookup);
    if (verseText == null) {
      unresolved.push(body.trim());
      return match;
    }
    // Wrap with blank lines so the blockquote always becomes its own paragraph,
    // even when the model emitted the sentinel mid-sentence. Trailing collapse
    // normalizes runs of >2 newlines back to a single paragraph break.
    return `\n\n> "${verseText}"\n> <cite>(${ref.viBook} ${ref.display})</cite>\n\n`;
  });
  const output = replaced.replace(/\n{3,}/g, "\n\n").trim();
  return { output, unresolved };
}
